hathor/indexes/rocksdb_deps_index.py

```diff
@@ -0,0 +1,329 @@
+        import struct
+        assert len(value) == 4 + 32
+        height, = struct.unpack('!I', value[:4])
+        tx_hash = value[4:]
+        return height, tx_hash
+
+    def _iter_rev_deps_of(self, tx_dep_hash: bytes) -> Iterator[bytes]:
+        it = self._db.iterkeys(self._cf)
+        seek_key = self._to_key_reverse(tx_dep_hash)
+        self.log.debug('seek to start', seek_key=seek_key.hex())
+        it.seek(seek_key)
+        for _, key in it:
+            key_any = self._from_key_any(key)
+            if key_any.tag is not _Tag.REVERSE:
+                break
+            if key_any.tx_dep_hash != tx_dep_hash:
+                break
+            tx_hash = key_any.tx_hash
+            assert tx_hash is not None
+            self.log.debug('seek found', tx=tx_hash.hex())
+            yield tx_hash
+        self.log.debug('seek end')
+
+    def _del_from_deps(self, tx: BaseTransaction, batch: 'rocksdb.WriteBatch') -> None:
+        assert tx.hash is not None
+        for tx_dep_hash in tx.get_all_dependencies():
+            batch.delete((self._cf, self._to_key_reverse(tx_dep_hash, tx.hash)))
+
+    def _add_ready(self, tx_hash: bytes, batch: 'rocksdb.WriteBatch') -> None:
+        key = self._to_key_ready(tx_hash)
+        batch.put((self._cf, key), b'')
+
+    def add_tx(self, tx: BaseTransaction, partial: bool = True) -> None:
+        import rocksdb
+        assert tx.hash is not None
+        assert tx.storage is not None
+        batch = rocksdb.WriteBatch()
+        validation = tx.get_metadata().validation
+        if validation.is_fully_connected():
+            self._del_from_deps(tx, batch)
+        elif not partial:
+            raise ValueError('partial=False will only accept fully connected transactions')
+        else:
+            self._add_deps(tx, batch)
+            self._add_needed(tx, batch)
+        self._db.write(batch)
+
+    def del_tx(self, tx: BaseTransaction) -> None:
+        import rocksdb
+        batch = rocksdb.WriteBatch()
+        self._del_from_deps(tx, batch)
+        self._db.write(batch)
+
+    def _add_deps(self, tx: BaseTransaction, batch: 'rocksdb.WriteBatch') -> None:
+        assert tx.hash is not None
+        for dep in tx.get_all_dependencies():
+            batch.put((self._cf, self._to_key_reverse(dep, tx.hash)), b'')
+
+    def _add_needed(self, tx: BaseTransaction, batch: 'rocksdb.WriteBatch') -> None:
+        assert tx.hash is not None
+        assert tx.storage is not None
+        tx_storage = tx.storage
+
+        height = get_requested_from_height(tx)
+        self.log.debug('add needed deps', tx=tx.hash_hex, height=height, type=type(tx).__name__)
+        # get_tx_parents is used instead of get_tx_dependencies because the remote will traverse the parent tree, not
+        # the dependency tree, eventually we should receive all tx dependencies and be able to validate this tx
+        for tx_dep_hash in tx.get_tx_parents():
```

Comment:
I guess you changed it on the memory implementation. You are using the `get_tx_dependencies()`.

Reply:
You're right, I'll fix it.

Reply:
Done.

Reply:
Great! Did you check if any test was failing because of this problem?

Reply:
No tests fail because of this. I still don't have a clear picture of the specific case that would cause a problem, I just believe there might be one.

Reply:
Got it. Let's keep an open issue about it so we can get back to it later. I don't feel it's critical but I think we should have it properly tested (even if it means that we can change back to `get_tx_parents()`).

Reply:
I think I have a scenario where `get_tx_parents()` would fail.

The solid lines are for parents, and the dotted lines are for inputs.

<img width="714" alt="image" src="https://user-images.githubusercontent.com/449119/168301594-8bd504ac-310f-4641-ad35-1d6859636805.png">

```
txA.inputs = [txC]
txB.parents = [txA, txC]
```

In this case, txA would never have all dependencies if we don't include txC in the needed list.

Reply:
That's not exactly the case that will trigger the problem, but it's very very close. In your case `txC` _will_ be pulled when we get the parents from `txB`, which in the context of sync-v2 we go through `txB` _before_ going through `txA`. But if `txC` is a tip, then we have the issue. In the context of sync-v2, if `txC` is a tip, and there is no block that confirms it, then we would not try to fetch it by only using the parents.

Reply:
Hum got it. Nice catch! What if the block confirms A instead of B? In this case, B would be in the mempool.

Reply:
Hmm, yes, if there is no block confirming B, then the block-sync state of sync-v2 (which has to finish first before we start syncing the mempool) will not pull B, so that would in fact lead to the problem.

Reply:
Just pushed it, it got a little bigger than I imagined it would be. Either we really could use some ergonomic improvements to make these, or I'm really out of touch with how to use the ergonomics we have.

Reply:
@jansegre @msbrogli 

We have a real-life example of this situation on the mainnet:

[This transaction](https://explorer.hathor.network/transaction/000000008cd231281dc44c7c7ba8acc4da95848211bb6ba6cd67e45b538fa87f) is confirmed by a block with height [1727402](https://explorer.hathor.network/transaction/00000000000000001936b86fe501adcf29df93c5b0324cff0d432c5179705b4c) and uses funds from [a transaction](https://explorer.hathor.network/transaction/0000000090db1a72b14d7af871f72345689de27739bc374e44f724ba3d121146) that is confirmed by a block with height [1727403](https://explorer.hathor.network/transaction/00000000000000003902ef2592e5d30712c4b249f1eb5623360fc9cd8dfa54b3)